source/blog/2017-08-04-ovirt-webadmin-gwt-debug-quick-refresh.html.md

```diff
@@ -0,0 +1,132 @@
+---
+title: oVirt Webadmin GWT Debug Quick Refresh
+author: gshereme
+tags: GWT, webadmin, debugging
+date: 2017-08-04 15:00:00 CET
+comments: true
+published: true
+---
+
+As a developer, one drawback of using GWT for the oVirt Admin Portal (aka webadmin) is that the GWT compile process takes an exceptionally long time. If you make a change in some code and rebuild the ovirt-engine project using `make install-dev ...`, you'll be waiting several minutes to test your change. In practice, such a long delay in the usual code-compile-refresh-test cycle would be unbearable.
+
+READMORE
+
+Luckily, we can use GWT Super Dev Mode ("SDM") to start up a quick refresh-capable instance of the application. With SDM running, you can make a change in GWT and test the refreshed change within seconds.
+
+If you want to step through code and use the Chrome debugger, oVirt and SDM don't work well together for debugging due to Admin Portal's code and source map size. Therefore, we demonstrate how to disable source maps below.
+
+## Demo (40 seconds)
+
+<a href="http://www.youtube.com/watch?feature=player_embedded&v=IRGxIrwfnlo
+    " target="_blank"><img src="http://img.youtube.com/vi/IRGxIrwfnlo/0.jpg"
+    alt="demo" width="480" height="360" border="10" /></a>
+
+## Steps
+
+1. Open a terminal, build the engine normally, and start it.
+
+    ```
+    make clean install-dev PREFIX=$HOME/ovirt-engine DEV_EXTRA_BUILD_FLAGS_GWT_DEFAULTS="-Dgwt.cssResourceStyle=pretty -Dgwt.userAgent=safari" BUILD_UT=0 DEV_EXTRA_BUILD_FLAGS="-Dgwt.compiler.localWorkers=1"
+
+    ...
+
+    $HOME/ovirt-engine/share/ovirt-engine/services/ovirt-engine/ovirt-engine.py start
+
+    ```
+
+    ![screen](../images/blog/2017-08-04/1.png)
+
+    ![screen](../images/blog/2017-08-04/2.png)
+
+    ![screen](../images/blog/2017-08-04/3.png)
+
+2. In a second terminal, run:
+
+    Chrome:
+
+    ```
+    make gwt-debug DEV_BUILD_GWT_SUPER_DEV_MODE=1 DEV_EXTRA_BUILD_FLAGS_GWT_DEFAULTS="-Dgwt.userAgent=safari"
+    ```
+
+    or
+
+    Firefox:
+
+    ```
+    make gwt-debug DEV_BUILD_GWT_SUPER_DEV_MODE=1 DEV_EXTRA_BUILD_FLAGS_GWT_DEFAULTS="-Dgwt.userAgent=gecko1_8"
+    ```
+
+    ![screen](../images/blog/2017-08-04/4.png)
+
+    Wait about 2 mins for it to say "code server running at http://<<server>>:9876/", and the GWT Dev Mode app will open.
+
+    ![screen](../images/blog/2017-08-04/5.png)
+
+3.  Open your browser and go to http://localhost:9876/ **You must use localhost. It breaks with IPs, not sure why**
+
+    Now you'll see two buttons, Dev Mode On and Dev Mode Off.
+
+    ![screen](../images/blog/2017-08-04/6.png)
+
+    Drag those buttons to the browser bookmark bar, as it says on that page.
+
+    ![screen](../images/blog/2017-08-04/7.png)
+
+4. **Caveat** -- Ignore the link under the two buttons. Instead, in a second tab, go to webadmin via normal URL, http://localhost:8080
```

Comment:
Caveat refers to some type of condition. This appears to be a simple instruction. 
Therefore I suggest:
"4. Ignore the link under the two buttons. Instead, open another browser window and enter http://localhost:8080 in the address bar to go to webadmin."